INT-2460: Remove message modification logic in MS

artembilan opened the pull request:
JIRA: https://jira.spring.io/browse/INT-2460,
https://jira.spring.io/browse/INT-4122

Since the main purpose of the `MessageStore` to persist message for durability and only,
it doesn't make sense to modify `Message` for additional headers like `SAVED` and `CREATED_DATE`.
Such a logic should be a part of metadata stored together with the message.
And it is provided by the out-of-the-box `MessageStore` implementation.
In addition we free ourselves from the reflection operations to retain `ID` and `TIMESTAMP` headers when we add `SAVED` and `CREATED_DATE`
- Control "already saved" logic in the `JdbcMessageStore`s via `DuplicateKeyException` on the `INSERT`.
  This is much effective then additional `SELECT` in case of `SAVED` before
- Control "already saved" logic in the  `AbstractConfigurableMongoDbMessageStore` via `DuplicateKeyException` on the `INSERT`.
  Since `MongoDbMessageStore` doesn't provide extra `messageId` field, perform extra `SELECT` before store document.
  Anyway the `MongoDbMessageStore` isn't recommended for use.
  We may consider to deprecate it
- Control "already saved" logic  in the `AbstractKeyValueMessageStore` via `putIfAbsent` operation

With this fix we persist message in the store as is without any modifications when we perform standard serialization procedure.
Any custom serializers should consider to use `MutableMessageBuilder` if there is a requirement to retain `ID` and `TIMESTAMP`

artembilan committed:
INT-2460: Remove message modification logic in MS

JIRA: https://jira.spring.io/browse/INT-2460,
https://jira.spring.io/browse/INT-4122

Since the main purpose of the `MessageStore` to persist message for durability and only,
it doesn't make sense to modify `Message` for additional headers like `SAVED` and `CREATED_DATE`.
Such a logic should be a part of metadata stored together with the message.
And it is provided by the out-of-the-box `MessageStore` implementation.
In addition we free ourselves from the reflection operations to retain `ID` and `TIMESTAMP` headers when we add `SAVED` and `CREATED_DATE`

* Control "already saved" logic in the `JdbcMessageStore`s via `DuplicateKeyException` on the `INSERT`.
This is much effective then additional `SELECT` in case of `SAVED` before
* Control "already saved" logic in the  `AbstractConfigurableMongoDbMessageStore` via `DuplicateKeyException` on the `INSERT`.
Since `MongoDbMessageStore` doesn't provide extra `messageId` field, perform extra `SELECT` before store document.
Anyway the `MongoDbMessageStore` isn't recommended for use.
We may consider to deprecate it
* Control "already saved" logic  in the `AbstractKeyValueMessageStore` via `putIfAbsent` operation

With this fix we persist message in the store as is without any modifications when we perform standard serialization procedure.
Any custom serializers should consider to use `MutableMessageBuilder` if there is a requirement to retain `ID` and `TIMESTAMP`

artembilan commented:
This may require some Migration Guide, but let's see if the solution has right to go!

artembilan committed:
Rework `MongoDbMetadataStore.putIfAbsent()` to normal `findAndModify()` with particular `$setOnInsert`.
Technically the MongoDB query looks like:
```
db.collection.findAndModify({
  query: { _id: $key },
  update: {
    $setOnInsert: { value: $value } // perform modification only on upsert
  },
  new: false,   // don't return new doc if one is upserted
  upsert: true // insert the document if it does not exist
})
```

garyrussell commented:
:spring-integration-mongodb:checkstyleMain[ant:checkstyle] [ERROR] /home/travis/build/spring-projects/spring-integration/spring-integration-mongodb/src/main/java/org/springframework/integration/mongodb/metadata/MongoDbMetadataStore.java:23:8: Unused import - org.springframework.data.mongodb.core.CollectionCallback. [UnusedImports]

artembilan committed:
Move single import to the appropriate JavaDoc

garyrussell commented on spring-integration-gemfire/src/main/java/org/springframework/integration/gemfire/store/GemfireMessageStore.java:
```diff
@@ -107,6 +107,13 @@ protected void doStore(Object id, Object objectToStore) {
+	protected void doStoreIfAbsent(Object id, Object objectToStore) {
+		Assert.notNull(id, "'id' must not be null");
+		Assert.notNull(objectToStore, "'objectToStore' must not be null");
+		this.messageStoreRegion.putIfAbsent(id, objectToStore);
```
Debug log if `null` not returned.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:
```diff
@@ -44,26 +41,32 @@
+	 * @deprecated since 5.0. This constant isn't used any more.
+	 */
+	@Deprecated
 	protected static final String CREATED_DATE = "CREATED_DATE";
```
Perhaps we should store a wrapper including this info - the date the message was stored might be important metadata.

We already have it in a column in JDBC.

Perhaps a new method `getMessageMetadata(id)`.

artembilan replied:
Agreed. The same we have with the `MessageDocument` for MongoDB.
That should be enough easy since it is key-value store without restrictions for the target `value`.

artembilan committed:
Polishing after rebase
DEBUG messages in `doStoreIfAbsent()` implementations

artembilan committed:
* To keep track of the extra message information in the `MessageStore`, without `Message` modification, introduce `MessageMetadata` and `MessageHolder`
* Add `MessageStore#getMessageMetadata()`
* Modify MongoDb `MessageStore` to add extra `timestamp` for individual message
* Fix `ConcurrentAggregatorTests` race condition.
Since currently the default release strategy is `SimpleSequenceSizeReleaseStrategy` which is just based on the `MessageGroup` size, there is no guaranty which messages will complete the group in concurrent environment.
The test is really based on the `SequenceAwareMessageGroup` logic to discard the message with the same `correlationId`

artembilan committed:
Fix `@Copyright` format

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:
```diff
@@ -44,39 +41,61 @@
+		Assert.notNull(messageId, "'messageId' must not be null");
+		MessageHolder messageHolder = (MessageHolder) doRetrieve(MESSAGE_KEY_PREFIX + messageId);
+		if (messageHolder != null) {
+			Assert.isInstanceOf(MessageHolder.class, messageHolder);
```
? We would have gotten a `ClassCastException` above.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:
```diff
@@ -44,39 +41,61 @@
+		Assert.notNull(messageId, "'messageId' must not be null");
+		MessageHolder messageHolder = (MessageHolder) doRetrieve(MESSAGE_KEY_PREFIX + messageId);
+		if (messageHolder != null) {
+			Assert.isInstanceOf(MessageHolder.class, messageHolder);
```
Ditto

artembilan committed:
Move cast to `MessageHolder` after `Assert.isInstanceOf(MessageHolder.class, messageHolder)`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/AbstractKeyValueMessageStore.java:
```diff
@@ -44,39 +41,61 @@
+		Assert.notNull(messageId, "'messageId' must not be null");
+		Object messageHolder = doRetrieve(MESSAGE_KEY_PREFIX + messageId);
+		if (messageHolder != null) {
+			Assert.isInstanceOf(MessageHolder.class, messageHolder);
```
One last though, perhaps we can support `Message<?>` here - it would provide backwards compatibility - messages stored with old versions can still be retrieved.

WDYT?

artembilan replied:
M-m-m. Do you mean no `MessageHolder`, but `Message<?>` extension-wrapper, like in the `MessageWithThreadState`? In other words one more `MessageDecorator`

garyrussell replied:
No, I mean that with 4.x, `doStore()` stores a `Message<?>` in 5.0, it will store a `MessageHolder`.

We could support `doRetrieve()` returning old `Message<?>` objects rather than insisting they are holders.

This way, someone upgrading from 4.x to 5.0 won't have to wipe out the message store, over time the old `Message<?>`s will be removed and eventually the store will only have holders.

Obviously, the new `getMessageMetadata()` will only work with holders.

artembilan replied:
OK. That's make sense as well. Although with `MessageHolder<?> implements Message<?>` we kill both birds and only the `null` in the `getMessageMetadata()` if the result isn't `MessageHolder`.
However we can use deprecated headers to extract `MessageMetadata` for old messages.
WDYT?

garyrussell replied:
I don't think we need to support `getMessageMetadata()` for old records - I suppose returning null would be more friendly than throwing an illegal argument exception, though.

artembilan committed:
Retain backward compatibility in the `AbstractKeyValueMessageStore`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/store/MessageStore.java:
```diff
@@ -39,6 +40,15 @@
+	 * if no Message with that id exists in the MessageStore.
+	 * @since 5.0
+	 */
+	MessageMetadata getMessageMetadata(UUID id);
```
Perhaps...
```
/**
 * Return a {@link MessageMetadata} for the {@link Message} by provided {@code id}.
 * @param id The message identifier.
 * @return The MessageMetadata with the given id, or <i>null</i>
 * if no Message with that id exists in the MessageStore
 * or the message has no metadata (legacy message from an earlier version).
 * @since 5.0
 */
```

If you agree, I can fix on merge.

artembilan replied:
:+1:

garyrussell commented:
Merged as a1f554c04d956d029230c53ad0bf2e01c0354491
